enhancements/api-review/azure_user_defined_tags.md

```diff
@@ -0,0 +1,322 @@
+have the ability to apply user defined tags to those resources created by Openshift 
+in Azure.
+
+## Motivation
+
+Motivations include but are not limited to:
+
+- Allow admin, compliance, and security teams to keep track of assets and objects 
+  created by OpenShift in Azure.
+
+### User Stories
+
+- As an openshift administrator, I want to have tags added to all resources created 
+  in Azure by Openshift.
+- As an openshift administrator, I want to restrict access granted to Openshift 
+  specific account.
+
+### Goals
+
+- The administrator or service (in the case of Managed OpenShift) installing OpenShift 
+  can configure allowed number of user-defined tags in the Openshift installer generated
+  install config, which is referred and applied by the installer and the in-cluster operators
+  on the the Azure resources during cluster creation.
+- Tags must be applied at creation time, in an atomic operation. It isn't acceptable 
+  to create an object and to apply tags post cluster creation.
+
+### Non-Goals
+
+- Management(update/delete) of resource tags post creation of cluster is out of scope.
+
+## Proposal
+
+A tag of the form `kubernetes.io/cluster/<cluster_name>:owned` will be added to every
+resource created by Openshift to enable administrator to differentiate the resources
+created for Openshift cluster. An administrator is not allowed to add or modify the tag 
+having the prefix `kubernetes.io` or `openshift.io` in the name. 
+
+New `userTags` field will be added to `platform.azure` of install-config for the user 
+to define the tags to be added to the resources created by installer and in-cluster operators.
+
+If `platform.azure.userTags` of install-config has any tag defined same will be added 
+to all the azure resources created by Openshift except, when the tag validation fail
+to meet any of the below conditions
+1. A tag name can have a maximum of 128 characters.
+    - Tag name has a limit of 512 characters for all resources except for 
+    storage accounts, which has a limit of 128 characters and hence tag name 
+    length is restricted to 128 characters on every resource required by Openshift.
+2. A tag name cannot contain `<, >, %, &, \, ?, /, #, :, whitespace` characters and 
+   must not start with a number.
+    - DNS zones, Traffic, Front Door resources does not support tag with spaces, 
+    special/unicode characters or starting with number, hence these are added as 
+    constraints on every other Azure resource required by Openshift as well.
+3. A tag value can have a maximum of 256 characters.
+4. A resource, resource-group or subscription, user can configure a maximum of 5 tags
+   through Openshift. 
+    - Azure supports a maximum of 50 tags except for Automation, Content Delivery Network,
+    DNS resources which can have a maximum of 15 tags, hence restricting the number of 
+    user defined tags to 5 and 10 for Openshift's internal use, for all the resources 
+    created by Openshift.
```

Comment:
we recently had to bump up the number of user tags allowed for AWS resources, are we convinced 5 user tags will be enough for azure customers?  Do we really need to reserve 10 for openshift?

Reply:
Yeah, I guess Azure users too won't be happy with the limit. In Azure, out of the resources OCP requires, just DNS has the limitation and is created by installer and no operator creates new. 
Would it be acceptable to consider the first 10 userTags for DNS resource alone without throwing error, and have the limitation of 40 tags for other resources, with 10 reserved for OCP?
During one of the discussion, what I learnt was, 4 OCP specific tags are added to AWS resources by SRE team, and thought 10 would be a safer value, in case OCP needs to add few more tags in future.

Reply:
> 4 OCP specific tags are added to AWS resources by SRE team

Wouldn't SRE add those via this mechanism(setting the tags in spec.userTags), so they'd be part of the 5 user tag limit, not the 10 openshift tag limit?

Also aren't there really 3 categories of tag here?

1) owned/generated by openshift for its own purposes (you say this is limited to 10)
2) specified by the user in spec.userTags, but applied by openshift controllers to the resource (you say this is limited to 5)
3) specified by the user directly via the cloud interface (this is effectively limited to "whatever is left")

When you say "5 are reserved for user tags" is that category 2, or category 3?

Reply:
> Wouldn't SRE add those via this mechanism(setting the tags in spec.userTags), so they'd be part of the 5 user tag limit, not the 10 openshift tag limit?

Yeah, with just 5 allowed for user configuration, and 4 is known to be used, it just leaves out 1 for the user and I thought to get the optimal split up during the review.

> When you say "5 are reserved for user tags" is that category 2, or category 3?

5 reserved tags for the user belongs to category 2.

Reply:
We're only using one tag in category 1, right?  (the `owned` tag).  So do we really need to reserve 10 tags for that category?

Reply:
Category 1 has just one tag. Except for aws-load-balancer-operator, which adds an additional tag, but that's in AWS. Did not find any operator managing Azure resources adding additional tags other than `kubernetes.io.cluster:owned` tag.

Yeah, 10 won't be required, I think 5 for OCP should be sufficient.

Reply:
> Yeah, 10 won't be required, I think 5 for OCP should be sufficient.

so should we flip this, reserve 5 for OCP and limit the user to 10?

Reply:
Yeah, unless someone objects with a requirement for more tags :)
Changes updated.

Reply:
As SRE team makes use of userTags for adding tags and tags reserved for OCP are for those added internally by installer & operators like the `kubernetes.io.cluster:owned` default tag, thought 5 should be enough. Do we need to reserve more for OCP's future use?

Reply:
> Can we enumerate the 5 that are in use here because we mention 2 in use for openshift and 4 in use by SRE, which would make 6, so are we sure 5 is enough?

we don't have 5 in use currently, we have 1 in use currently(kubernetes.io.cluster:owned)  (two if you count the aws loadbalancer scenario described above), as i understand it.

So 5 feels like a decent amount of headroom for our needs and we're already getting asked by users to reserve more tags for their usage, so swinging in the balance in that direction seems justified.

Reply:
For AWS, we take the first N (I believe it is 8) lexicographically sorted user tags and apply those to the resources that have lower limits than the cluster limits.

I am not sure if this would be acceptable in the case of "dns resources," but if we think customers are not too concerned about tagging this resource, it seems like it would allow more tags for the cluster.

Reply:
I think we should resolve this thread (i.e. update the enhancement) before moving forward. Are there any issues with an outcome of 50 tags (with a 25/25 user/openshift split)? 

DNS resources are the only cluster resources with the lower limit. Perhaps we could just skip tagging these? If not, we could apply an arbitrary or lexicographically sorted subset of the user provided tags. Keep in mind that both the installer and cluster ingress operator create dns records.

Reply:
can we get a resolution on whether we need to be tagging dns resources or not?

i don't have an issue w/ a 25/25 split for resources that can handle 50.  For resources that can only handle 15, i'd like to see 5(openshift)/10(user) and lexographic sorting so we get a deterministic behavior when truncating the list.  But if someone can get an actual PM answer on whether DNS resources *need* tagging, that would be even better.

Reply:
@patrickdillon @bparees @JoelSpeed Would it be acceptable if we proceed with 10(user)/5(openshift) split for the tech preview as pointed out by @TrilokGeer.

Reply:
> @patrickdillon @bparees @JoelSpeed Would it be acceptable if we proceed with 10(user)/5(openshift) split for the tech preview as pointed out by @TrilokGeer.

are you suggesting that as the limits for all resources, or the limits for resources that don't allow more tags?

Reply:
I would like to suggest to have the limitation of max 10 user tags for all the resources.

Reply:
> I would like to suggest to have the limitation of max 10 user tags for all the resources.

since it is easier to raise a limit later than lower it, i guess we can start there and then raise it for the other resources if needed (at the cost of more code complexity and need for clear documentation about the behavior for different resource types).